gateway/downlink.go

```
@@ -150,22 +154,24 @@ func (g *gateway) createDownlink(device *node.Node, framePayload []byte) ([]byte
 	// fopts := []byte{0x3A, 0xFF, 0x00, 0x01}
 
 	// payload = append(payload, fopts...)
-
-	payload = append(payload, device.FPort)
-
 	// TODO (om) commented for future testing
 	// 30 seconds
 	// framePayload := []byte{0x01, 0x00, 0x00, 0x1E} //  dragino
 	// framePayload := []byte{0xff, 0x10, 0xff} //tilt reset
 
-	encrypted, err := crypto.EncryptDownlink(
-		types.AES128Key(device.AppSKey), *types.MustDevAddr(device.Addr), device.FCntDown+1, framePayload)
-	if err != nil {
-		return nil, err
+	if framePayload != nil {
+		if device.FPort == 0 {
```

we should also check the FPort is non zero in `register_device do command`

---

we don't validate any of the other fields in the do command so it feels weird to me to do it there, I could add another check to the validate function to error if the user puts zero

---

that might be fine, big thing I want to avoid is a user suddenly seeing errors because their custom node is misconfigured somehow. super unlikely but would be better to surface the error sooner rather than at downlink time.